dartsim/src/EntityManagementFeatures.cc

```diff
@@ -304,25 +307,29 @@ Identity EntityManagementFeatures::GetLink(
 Identity EntityManagementFeatures::GetLink(
     const Identity &_modelID, const std::string &_linkName) const
 {
-  DartBodyNode *const bn =
-      this->ReferenceInterface<ModelInfo>(_modelID)->model->getBodyNode(
-          _linkName);
-
-  // If the link doesn't exist in "links", it means the containing entity has
-  // been removed.
-  if (this->links.HasEntity(bn))
-  {
-    const std::size_t linkID = this->links.IdentityOf(bn);
-    return this->GenerateIdentity(linkID, this->links.at(linkID));
-  }
-  else
+  auto modelInfo = this->ReferenceInterface<ModelInfo>(_modelID);
+  for (const auto &linkInfo : modelInfo->links)
   {
-    // TODO(addisu) It's not clear what to do when `GetLink` is called on a
-    // model that has been removed. Right now we are returning an invalid
-    // identity, but that could cause a segfault if the use doesn't check if
-    // returned value before using it.
-    return this->GenerateInvalidId();
+    if (_linkName == linkInfo->name)
+    {
+      // If the link doesn't exist in "links", it means the containing entity
+      // has been removed.
+      if (this->links.HasEntity(linkInfo->link))
+      {
+        const std::size_t linkID = this->links.IdentityOf(linkInfo->link);
+        return this->GenerateIdentity(linkID, this->links.at(linkID));
+      }
+      else
+      {
+        // TODO(addisu) It's not clear what to do when `GetLink` is called on a
+        // model that has been removed. Right now we are returning an invalid
+        // identity, but that could cause a segfault if the use doesn't check if
+        // returned value before using it.
+        return this->GenerateInvalidId();
```

Comment:
Would it at least be worth removing the link from `modelInfo->links` if we find the link has been removed here?

Reply:
While I agree that `modelInfo->links` and `Base::links` being out of sync causes an invalid state, I'd rather not modify `modelInfo->links` from a const function that's simply supposed to return a link identity. If this case actually occurs, I think we should fix the root cause, which is that the `RemoveModel` function is not working properly.